### PR TITLE
Example for Tensorflow 2.x and smdebug

### PR DESCRIPTION
**Pull request by vandanavk**
*Issue #, if available:*

*Description of changes:*
Adding examples for using TF 2.x fit API with SageMaker Debugger v0.7.1+.

Example 1: Modifying the training script to add sagemaker debugger hook, to be used with a custom container
Example 2: Using sagemaker debugger without modifying training script but configuring the debugger hook through SageMaker Estimator.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by vandanavk** [comment]
@Vikas-kum for review

**Comment by Vikas-kum** [comment]
<p>What is custom container vs default container?</p>

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+"""
+This script is a ResNet training script which uses Tensorflow's Keras interface.
+It has been orchestrated with SageMaker Debugger hooks to allow saving tensors during training.
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:3]: This script provides example of how to use sagemaker debugger if you use 
  1) your own custom container to run in sagemaker
  2) your own script outside sagemaker

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+These hooks have been instrumented to read from json configuration that SageMaker will put in the training container.
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:4]: this can be removed.

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+Configuration provided to the SageMaker python SDK when creating a job will be passed on to the hook.
+This allows you to use the same script with differing configurations across different runs.
+If you use an official SageMaker Framework container (i.e. AWS Deep Learning Container), then
+you do not have to orchestrate your script as below. Hooks will automatically be added in those environments.
+For more information, please refer to https://github.com/awslabs/sagemaker-debugger/blob/master/docs/sagemaker.md
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:9]: refer to right example notebook.

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
+"""
+This script is a ResNet training script which uses Tensorflow's Keras interface.
+It has been orchestrated with SageMaker Debugger hooks to allow saving tensors during training.
+These hooks have been instrumented to read from json configuration that SageMaker will put in the training container.
+Configuration provided to the SageMaker python SDK when creating a job will be passed on to the hook.
+This allows you to use the same script with differing configurations across different runs.
+If you use an official SageMaker Framework container (i.e. AWS Deep Learning Container), then
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:7]: one line gap

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+you do not have to orchestrate your script as below. Hooks will automatically be added in those environments.
+For more information, please refer to https://github.com/awslabs/sagemaker-debugger/blob/master/docs/sagemaker.md
+"""
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:10]: "For quickly navigating smdebug specific changes search for SMDEBUG " 
Then add # SMDEBUG comment to section where SMDEBUG specific code is there.

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    X_train = X_train.astype('float32')
+    X_valid = X_valid.astype('float32')
+
+    mean_image = np.mean(X_train, axis=0)
+    X_train -= mean_image
+    X_valid -= mean_image
+    X_train /= 128.
+    X_valid /= 128.
+
+    hook.set_mode(smd.modes.TRAIN)
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:40]: Explain why it is done.  
This is not mandatory. This step will help us to differentiate TRAINING mode with EVAL mode. To learn more about modes follow {Link to doc explaining modes. }

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    model.fit(X_train, Y_train,
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:41]: do we have to set mode explicitly in keras? 
We should get callback TRAIN & EVAL , no?

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    args = parser.parse_args()
+
+    if args.random_seed:
+        tf.random.set_seed(2)
+        np.random.seed(2)
+        random.seed(12)
+
+    model = ResNet50(weights=None, input_shape=(32,32,3), classes=10)
+
+    # Create hook from the configuration provided through sagemaker python sdk
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:77]: # SMDEBUG

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    hook = smd.KerasHook.create_from_json_file()
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:78]: comment here. 
"We will create a hook from json file config. Example of json file config: {} " 
If you are using sagemaker estimator sdk, you can provide config in estimator fit call. { link }

File: sagemaker-debugger/tensorflow2/tensorflow2_zero_code_change/src/keras_fit_zero_code_change.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
+"""
+This script is a ResNet training script which uses Tensorflow's keras fit() API.
+Using this script with the official SageMaker Framework container (i.e. AWS Deep Learning Container), enables you to make use of the SageMaker Debugger without any Debugger related
+additions to the training script.
+For more information, please refer to https://github.com/awslabs/sagemaker-debugger/blob/master/docs/sagemaker.md
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_zero_code_change/src/keras_fit_zero_code_change.py:5]: link about zero code change and what is saved by ZCC by default and what can be configured in estimator sdk(default collections , custom collections etc)

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/tf2-keras-custom-container.ipynb [hunk 11]
```diff
@@ -0,0 +1,319 @@
+{
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/tf2-keras-custom-container.ipynb:1]: <p>1)Full description of 2.x support is available at {link} -- on the link we should have table saying what is supported in TF.115 | TF.2.0,TF2.1</p><p> </p><p>2) Amazon SageMaker Debugger{link to debugger github} is a new feature</p><p> </p><p>3) In this notebook, we'll show you how to use a built-in rule to monitor your training job using a tf.keras (TF 2.1.0) ResNet example.  --&gt; we 'll show how to use sagemaker-debugger and use a built-in rule {Rulename} to monitor training job in real time.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1102/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/tf2-keras-custom-container.ipynb:1]: <p>if user will run this it wont use custom container , rather default container? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1102/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

File: sagemaker-debugger/tensorflow2/tensorflow2_zero_code_change/tf2-keras-default-container.ipynb [hunk 12]
```diff
@@ -0,0 +1,396 @@
+{
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_zero_code_change/tf2-keras-default-container.ipynb:1]: <p>provide links to options available for default collections. </p><p> </p><p>Also provide links to how do users get to their own custom collection</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/1102/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py (hunk 6 above)
Comment (Vikas-kum) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:40]: And may be for keras, hook will automatically get the mode, right? You can right in comment that this is optional. 
Also will this line always cause tensors to be saved in TRAIN mode, even eval mode tensors would be saved in TRAIN mode?

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/tf2-keras-custom-container.ipynb [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "estimator = TensorFlow(\n",
+    "    role=sagemaker.get_execution_role(),\n",
+    "    base_job_name='smdebug-demo-tf2-keras',\n",
+    "    ## custom container with TF 2.1.0 and smdebug 0.7.1 installed\n",
+    "    #image_name='',\n",
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/tf2-keras-custom-container.ipynb:134]: may be more comment here and above, which tells that they have to insert their own image and link BYOC page. Actually information about BYOC should be at the top of notebook.

**Comment by vandanavk** [comment]
> What is custom container vs default container?

BYOC vs SM DLC

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py (hunk 1 above)
Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:4]: changed in the latest commit

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py (hunk 3 above)
Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py]: changed in the latest commit

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py (hunk 4 above)
Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:9]: done

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py (hunk 5 above)
Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:11]: done in the latest commit

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py (hunk 6 above)
Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py]: removed it as it is not mandatory. as you rightly pointed out, it gets set in the callbacks

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py (hunk 7 above)
Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:43]: no, removed this line

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py (hunk 8 above)
Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py]: done in the latest commit

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py (hunk 9 above)
Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/src/tf_keras_resnet_byoc.py:89]: added details in the latest commit

File: sagemaker-debugger/tensorflow2/tensorflow2_zero_code_change/src/keras_fit_zero_code_change.py (hunk 10 above)
Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_zero_code_change/src/keras_fit_zero_code_change.py]: done

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/tf2-keras-custom-container.ipynb (hunk 11 above)
Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/tf2-keras-custom-container.ipynb:1]: 2 and 3 -> done

1 -> we dont have public documentation about this. probably add this in smdebug docs? or the aws doc where the limitations of smdebug are mentioned

Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/tf2-keras-custom-container.ipynb:1]: correct. if user doesn't mention an image name, it will use SM DLC

File: sagemaker-debugger/tensorflow2/tensorflow2_zero_code_change/tf2-keras-default-container.ipynb (hunk 12 above)
Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_zero_code_change/tf2-keras-default-container.ipynb:1]: done

File: sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/tf2-keras-custom-container.ipynb (hunk 13 above)
Comment (vandanavk) [reply sagemaker-debugger/tensorflow2/tensorflow2_keras_custom_container/tf2-keras-custom-container.ipynb]: added a link about BYOC at the top of the notebook
